### DISCUSSION TITLE
How would function calls be differentiated from non-function calls?

### DISCUSSION THREAD

**Discussion by mybearworld** [TodePond/GulfOfMexico#424] [Q&A] [answered]
Consider this:

```java
u foo => bar!
u bar => foo!
```

Or more verbose:

```java
function foo () => bar()!
function bar () => foo()!
```

Is it recursive, or will it just output each other's value? How would you achieve the other thing?

**Comment by treeplate** [comment]
i think line 1 would turn bar() into a string because bar isn't defined yet

**Reply by vivaansinghvi07** [reply]
not necessarily, depending on the implementation, the program will check if `bar` is in the namespace only when `foo` is called

**Comment by circle-gon** [comment]
Very good point. From the language author's perspective, there isn't really anything we can do. From the programmer, it could be mitigated by inserting a `undefined` argument, so the interpreter can look at it correctly.

**Reply by vivaansinghvi07** [reply]
That’s actually exactly how I do it. When running a  through the code, any set of () is replaced with an empty Name token, which is interpreted as a `DreamberdSpecialBlankValue` and ignored when the function itself is called

**Reply by circle-gon** [reply]
I thought about that, but I thought it would violate the spec (since they aren't always replaced as whitespace), and that it would make parentheses actually have a function instead of being useless (also violating the spec).

**Reply by vivaansinghvi07** [reply]
That’s true, but I figured that violation is worth having to enforce putting at least one argument in every function call

**Comment by TodePond** [comment, marked as answer]
This would cause a runtime stack overflow crash.

```java
function foo () => bar()!
function bar () => foo()!

foo()!
```

As would this:

```java
u foo => bar!
u bar => foo!

foo!
```

**Reply by circle-gon** [reply]
So passing functions around by value no longer works? Since it seems like that any mention of a function will automatically invoke it.

**Reply by mybearworld** [reply]
Hmm, you could probably work around that with `c() => bar`. Thank you!

**Reply by TodePond** [reply]
> So passing functions around by value no longer works? Since it seems like that any mention of a function will automatically invoke it.

Only for zero-argument functions.
